gobblin-modules/gobblin-metadata/src/main/java/gobblin/metadata/provider/ConfigurableMetadataProviderFactory.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gobblin.metadata.provider;
+
+import com.typesafe.config.Config;
+
+import gobblin.util.ConfigUtils;
+import gobblin.util.reflection.GobblinConstructorUtils;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Configurable {@link DatasetAwareMetadataProviderFactory} that uses {@link #METADATA_PROVIDER_CLASS} and
+ * {@link #METADATA_PROVIDER_URI} to instantiate a {@link DatasetAwareMetadataProvider}.
+ */
+@Slf4j
+public class ConfigurableMetadataProviderFactory implements DatasetAwareMetadataProviderFactory {
```

Comment:
The comment was probably not clear. The pattern I suggested is:

```java
new ClassAliasResolver(DatasetAwareMetadataProviderFactory.class).resolveClass(userProvidedAlias).newInstance().createMetadataProvider(metaConfig);
```

Since the Factory constructor is guaranteed to have no arguments, `ConstructorUtils` is not needed. On the other hand, the factory itself does not need to use a `Class.forName()` or `ConstructorUtils` as the class has already been selected.